test/e2e/framework/kcp-ingress.go

```diff
@@ -0,0 +1,284 @@
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// kcpIngress exposes a kcp-ingress invocation to a test and
+// ensures the following semantics:
+//  - the server will run only until the test deadline
+//  - all ports and data directories are unique to support
+//    concurrent execution within a test case and across tests
+type kcpIngress struct {
+	args              []string
+	ctx               context.Context
+	artifactDir       string
+	kcpKubeconfigPath string
+
+	lock *sync.Mutex
+	cfg  *rest.Config
+
+	t TestingTInterface
+}
+
+func newKcpIngress(t *T, kcpKubeconfigPath, artifactDir string) (*kcpIngress, error) {
+	t.Helper()
+	ctx := context.Background()
+	if deadline, ok := t.Deadline(); ok {
+		if remaining := time.Until(deadline); remaining < 30*time.Second {
+			return nil, fmt.Errorf("only have %v until deadline, need at least 30 seconds", remaining)
+		}
+		c, cancel := context.WithDeadline(ctx, deadline.Add(-10*time.Second))
+		ctx = c
+		t.Cleanup(cancel) // this does not really matter but govet is upset
+	}
+	envoyListenerPort, err := GetFreePort(t)
+	if err != nil {
+		return nil, err
+	}
+	envoyXDSPort, err := GetFreePort(t)
+	if err != nil {
+		return nil, err
+	}
+
+	artifactDir = filepath.Join(artifactDir, "kcp-ingress")
+	if err := os.MkdirAll(artifactDir, 0755); err != nil {
+		return nil, fmt.Errorf("could not create artifact dir: %w", err)
+	}
+
+	return &kcpIngress{
+		args: append([]string{
+			"-envoyxds",
+			"-envoyxds-port=" + envoyXDSPort,
+			"-envoy-listener-port=" + envoyListenerPort,
+			"-kubeconfig=" + kcpKubeconfigPath},
+		),
+		artifactDir: artifactDir,
+		ctx:         ctx,
+		t:           t,
+		lock:        &sync.Mutex{},
+	}, nil
+}
+
+// Run runs the kcp server while the parent context is active. This call is not blocking,
+// callers should ensure that the server is Ready() before using it.
+func (c *kcpIngress) Run(parentCtx context.Context) error {
+	// calling any methods on *testing.T after the test is finished causes
+	// a panic, so we need to communicate to our cleanup routines when the
+	// test has been completed, and we need to communicate back up to the
+	// test when we're done with everything
+	ctx, cancel := context.WithCancel(parentCtx)
+	if deadline, ok := c.t.Deadline(); ok {
+		deadlinedCtx, deadlinedCancel := context.WithDeadline(ctx, deadline.Add(-10*time.Second))
+		ctx = deadlinedCtx
+		c.t.Cleanup(deadlinedCancel) // this does not really matter but govet is upset
+	}
+	c.ctx = ctx
+	cleanupCtx, cleanupCancel := context.WithCancel(context.Background())
+	c.t.Cleanup(func() {
+		c.t.Log("cleanup: ending kcp-ingress controller")
+		cancel()
+		<-cleanupCtx.Done()
+	})
```

Comment:
Possible follow-up issue: I wonder if we could extract this logic, as we're going to need it every time we add a new binary/struct like kcp & kcpIngress to the framework. cc @stevekuznetsov

Reply:
I've moved this PR to WIP as I'm adding the e2e tests... don't worry about this file "test/e2e/framework/kcp-ingress.go" yet. :)

Reply:
Also, If we go in this direction, we will need to change slightly the Run function, with some config interface or so:

https://github.com/kcp-dev/kcp/blob/main/test/e2e/framework/test.go#L70